ssl/quic/quic_port.c

```diff
@@ -647,9 +658,164 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
     return i > 0;
 }
 
-#define TOKEN_LEN (sizeof("openssltoken") + \
-                   QUIC_RETRY_INTEGRITY_TAG_LEN - 1 + \
-                   sizeof(unsigned char))
+static void cleanup_validation_token(QUIC_VALIDATION_TOKEN *token)
+{
+    OPENSSL_free(token->remote_addr);
+}
+
+/**
+ * @brief Generates a validation token for a RETRY packet.
+ *
+ * @param peer  Address of the client peer receiving the packet.
+ * @param odcid DCID of the connection attempt.
+ * @param rscid Retry source connection ID of the connection attempt.
+ * @param token Address of token to fill data.
+ *
+ * @return 1 if validation token is filled successfully, 0 otherwise.
+ */
+static int generate_retry_token(BIO_ADDR *peer, QUIC_CONN_ID odcid,
+                                QUIC_CONN_ID rscid, QUIC_VALIDATION_TOKEN *token)
+{
+    token->is_retry = 1;
+    token->timestamp = ossl_time_now();
+    token->remote_addr = NULL;
+    token->odcid = odcid;
+    token->rscid = rscid;
+
+    if (!BIO_ADDR_rawaddress(peer, NULL, &token->remote_addr_len)
+        || token->remote_addr_len == 0
+        || (token->remote_addr = OPENSSL_malloc(token->remote_addr_len)) == NULL
+        || !BIO_ADDR_rawaddress(peer, token->remote_addr,
+                                &token->remote_addr_len)) {
+        OPENSSL_free(token->remote_addr);
+        return 0;
+    }
+
+    return 1;
+}
+
+/**
+ * @brief Marshals a validation token into a new buffer.
+ *
+ * Dynamically allocates |buffer| and stores the size of |buffer| in BUFFER_LEN.
+ * The caller is responsible for freeing |buffer|.
+ *
+ * @param token      Validation token.
+ * @param buffer     Callee will allocate a buffer and store the address here.
+ * @param buffer_len Size of |buffer|.
+ */
+static int marshal_validation_token(QUIC_VALIDATION_TOKEN *token,
+                                    unsigned char **buffer, size_t *buffer_len)
+{
+    size_t offset = 0;
+    unsigned char *buf;
+
+    buf = OPENSSL_malloc(1 + sizeof(token->timestamp)
+                         + sizeof(token->remote_addr_len));
+    if (!buf)
+        return 0;
+
+    if (token->is_retry != 0 && token->is_retry != 1)
+        return 0;
+    memset(buf, token->is_retry, 1);
+    offset += 1;
+
+    memcpy(buf + offset, &token->timestamp, sizeof(token->timestamp));
+    offset += sizeof(token->timestamp);
+
+    memcpy(buf + offset, &token->remote_addr_len,
```

Comment:
do we really need size_t for remote_addr_len? My assumption is we deal with IP address (16B) and port number (2B), which is typically 18Bytes.

in case of dgram socket it can be kind of socket name which might MAX_PATH (?~4k?), so I think we can get away with 2B (16-bit for remote_addr_len).

On the other hand I'm not concerned that much now as marshalling will be using ASN1 or something like that. So I don't insist o this change now.

Reply:
Ok, I will leave this as is since we will be using ASN1 at some point